File: sdk-typescript/packages/phone/src/phone.ts
/**
 * WAVE SDK - Phone API
 *
 * Voice calling and telephony integration capabilities.
 *
 * NOTE: This is a client SDK. All authorization checks are performed server-side.
 * The API will return 403 Forbidden if the user lacks required permissions.
 */

import type {
  WaveClient,
  PaginationParams,
  PaginatedResponse,
  Metadata,
} from '@wave-av/core';
import type { CallStatus, PhoneNumberType, PhoneNumber, Call, MakeCallRequest, UpdateCallRequest, Conference, ConferenceParticipant, SearchNumbersRequest, AvailablePhoneNumber, ListCallsParams } from './phone-types';
export type { CallStatus, CallDirection, PhoneNumberType, PhoneNumberCapabilities, PhoneNumber, Call, MakeCallRequest, UpdateCallRequest, Conference, ConferenceParticipant, SearchNumbersRequest, AvailablePhoneNumber, ListCallsParams } from './phone-types';

// Types

/**
 * Call status
 */

/**
 * Call direction
 */

/**
 * Phone number type
 */

/**
 * Phone number capabilities
 */

/**
 * Phone number
 */

/**
 * Call record
 */

/**
 * Make call request
 */

/**
 * Call update request
 */

/**
 * Conference room
 */

/**
 * Conference participant
 */

/**
 * Search available numbers request
 */

/**
 * Available phone number
 */

/**
 * List calls params
 */

// Phone API

/**
 * Phone API client
 */
export class PhoneAPI {
  private readonly client: WaveClient;
  private readonly basePath = '/v1/phone';

  constructor(client: WaveClient) {
    this.client = client;
  }

  /**
   * List owned phone numbers
   *
   * Requires: phone:read permission
   */
  async listNumbers(
    params?: PaginationParams & {
      status?: 'active' | 'inactive';
      type?: PhoneNumberType;
    }
  ): Promise<PaginatedResponse<PhoneNumber>> {
    return this.client.get<PaginatedResponse<PhoneNumber>>(
      `${this.basePath}/numbers`,
      { params: params as Record<string, string | number | boolean | undefined> }
    );
  }

  /**
   * Get a phone number by ID
   *
   * Requires: phone:read permission
   */
  async getNumber(numberId: string): Promise<PhoneNumber> {
    return this.client.get<PhoneNumber>(`${this.basePath}/numbers/${numberId}`);
  }

  /**
   * Search for available phone numbers to purchase
   *
   * Requires: phone:read permission
   */
  async searchAvailableNumbers(
    request: SearchNumbersRequest
  ): Promise<AvailablePhoneNumber[]> {
    return this.client.post<AvailablePhoneNumber[]>(
      `${this.basePath}/numbers/available`,
      request
    );
  }

  /**
   * Purchase a phone number
   *
   * Requires: phone:purchase permission
   */
  async purchaseNumber(
    number: string,
    options?: { friendly_name?: string; metadata?: Metadata }
  ): Promise<PhoneNumber> {
    return this.client.post<PhoneNumber>(`${this.basePath}/numbers/purchase`, {
      number,
      ...options,
    });
  }

  /**
   * Update a phone number
   *
   * Requires: phone:update permission
   */
  async updateNumber(
    numberId: string,
    updates: { friendly_name?: string; metadata?: Metadata }
  ): Promise<PhoneNumber> {
    return this.client.patch<PhoneNumber>(
      `${this.basePath}/numbers/${numberId}`,
      updates
    );
  }

  /**
   * Release a phone number
   *
   * Requires: phone:release permission (server-side RBAC enforced)
   */
  async releaseNumber(numberId: string): Promise<void> {
    await this.client.delete(
      `${this.basePath}/numbers/${numberId}`,
      { method: 'DELETE' }
    );
  }

  /**
   * Make an outbound call
   *
   * Requires: phone:call permission
   */
  async makeCall(request: MakeCallRequest): Promise<Call> {
    return this.client.post<Call>(`${this.basePath}/calls`, request);
  }

  /**
   * Get a call by ID
   *
   * Requires: phone:read permission
   */
  async getCall(callId: string): Promise<Call> {
    return this.client.get<Call>(`${this.basePath}/calls/${callId}`);
  }

  /**
   * List calls
   *
   * Requires: phone:read permission
   */
  async listCalls(params?: ListCallsParams): Promise<PaginatedResponse<Call>> {
    return this.client.get<PaginatedResponse<Call>>(
      `${this.basePath}/calls`,
      { params: params as Record<string, string | number | boolean | undefined> }
    );
  }

  /**
   * Update an active call
   *
   * Requires: phone:call permission
   */
  async updateCall(callId: string, updates: UpdateCallRequest): Promise<Call> {
    return this.client.patch<Call>(`${this.basePath}/calls/${callId}`, updates);
  }

  /**
   * End an active call
   *
   * Requires: phone:call permission
   */
  async endCall(callId: string): Promise<Call> {
    return this.updateCall(callId, { status: 'completed' });
  }

  /**
   * Get call recording
   *
   * Requires: phone:read permission
   */
  async getRecording(
    callId: string
  ): Promise<{ url: string; duration: number; file_size: number }> {
    return this.client.get(`${this.basePath}/calls/${callId}/recording`);
  }

  /**
   * Wait for call to end
   */
  async waitForCallEnd(
    callId: string,
    options?: {
      pollInterval?: number;
      timeout?: number;
      onUpdate?: (call: Call) => void;
    }
  ): Promise<Call> {
    const pollInterval = options?.pollInterval || 2000;
    const timeout = options?.timeout || 3600000; // 1 hour
    const startTime = Date.now();

    const terminalStatuses: CallStatus[] = [
      'completed',
      'failed',
      'busy',
      'no_answer',
      'canceled',
    ];

    while (Date.now() - startTime < timeout) {
      const call = await this.getCall(callId);

      if (options?.onUpdate) {
        options.onUpdate(call);
      }

      if (terminalStatuses.includes(call.status)) {
        return call;
      }

      await new Promise((resolve) => setTimeout(resolve, pollInterval));
    }

    throw new Error(`Call wait timed out after ${timeout}ms`);
  }

  /**
   * Create a conference room
   *
   * Requires: phone:conference permission
   */
  async createConference(options: {
    friendly_name: string;
    max_participants?: number;
    recording_enabled?: boolean;
    region?: string;
    metadata?: Metadata;
  }): Promise<Conference> {
    return this.client.post<Conference>(`${this.basePath}/conferences`, options);
  }

  /**
   * Get a conference by ID
   *
   * Requires: phone:read permission
   */
  async getConference(conferenceId: string): Promise<Conference> {
    return this.client.get<Conference>(
      `${this.basePath}/conferences/${conferenceId}`
    );
  }

  /**
   * List conferences
   *
   * Requires: phone:read permission
   */
  async listConferences(
    params?: PaginationParams & { status?: 'init' | 'in_progress' | 'completed' }
  ): Promise<PaginatedResponse<Conference>> {
    return this.client.get<PaginatedResponse<Conference>>(
      `${this.basePath}/conferences`,
      { params: params as Record<string, string | number | boolean | undefined> }
    );
  }

  /**
   * Add a participant to a conference
   *
   * Requires: phone:conference permission
   */
  async addConferenceParticipant(
    conferenceId: string,
    options: {
      from: string;
      to: string;
      muted?: boolean;
    }
  ): Promise<ConferenceParticipant> {
    return this.client.post<ConferenceParticipant>(
      `${this.basePath}/conferences/${conferenceId}/participants`,
      options
    );
  }

  /**
   * Update a conference participant
   *
   * Requires: phone:conference permission
   */
  async updateConferenceParticipant(
    conferenceId: string,
    callId: string,
    updates: { muted?: boolean; hold?: boolean }
  ): Promise<ConferenceParticipant> {
    return this.client.patch<ConferenceParticipant>(
      `${this.basePath}/conferences/${conferenceId}/participants/${callId}`,
      updates
    );
  }

  /**
   * Remove a participant from a conference
   *
   * Requires: phone:conference permission (server-side RBAC enforced)
   */
  async removeConferenceParticipant(
    conferenceId: string,
    callId: string
  ): Promise<void> {
    await this.client.delete(
      `${this.basePath}/conferences/${conferenceId}/participants/${callId}`,
      { method: 'DELETE' }
    );
  }

  /**
   * End a conference
   *
   * Requires: phone:conference permission
   */
  async endConference(conferenceId: string): Promise<Conference> {
    return this.client.post<Conference>(
      `${this.basePath}/conferences/${conferenceId}/end`
    );
  }

  /**
   * Validate a phone number
   *
   * Requires: phone:read permission
   */
  async validateNumber(
    number: string
  ): Promise<{
    valid: boolean;
    formatted_number?: string;
    country_code?: string;
    type?: PhoneNumberType;
    carrier?: string;
  }> {
    return this.client.post(`${this.basePath}/validate`, { number });
  }

  /**
   * Get supported countries
   *
   * Requires: phone:read permission
   */
  async getSupportedCountries(): Promise<
    Array<{
      code: string;
      name: string;
      calling_code: string;
      supported_types: PhoneNumberType[];
    }>
  > {
    return this.client.get(`${this.basePath}/countries`);
  }
}

/**
 * Create a Phone API instance
 */
export function createPhoneAPI(client: WaveClient): PhoneAPI {
  return new PhoneAPI(client);
}

File: sdk-python/wave/phone.py
"""WAVE SDK - Phone API. Voice calls, conferences, numbers, and recordings."""
from __future__ import annotations
import time
from typing import Any
from pydantic import BaseModel
from wave.client import WaveClient

class PhoneNumber(BaseModel):
    id: str; number: str; type: str; capabilities: list[str] | None = None; status: str; region: str | None = None; created_at: str; updated_at: str

class Call(BaseModel):
    id: str; from_number: str; to_number: str; status: str; direction: str; duration: float | None = None; recording_url: str | None = None; created_at: str

class Conference(BaseModel):
    id: str; name: str; status: str; participant_count: int = 0; max_participants: int = 100; created_at: str

class ConferenceParticipant(BaseModel):
    id: str; call_id: str; status: str; muted: bool = False; hold: bool = False; joined_at: str

class PhoneAPI:
    def __init__(self, client: WaveClient): self._client = client; self._base = "/v1/phone"
    def list_numbers(self, **params: Any) -> dict: return self._client.get(f"{self._base}/numbers", params={k: v for k, v in params.items() if v is not None})
    def get_number(self, number_id: str) -> PhoneNumber: return PhoneNumber(**self._client.get(f"{self._base}/numbers/{number_id}"))
    def purchase_number(self, number: str, **kwargs: Any) -> PhoneNumber: return PhoneNumber(**self._client.post(f"{self._base}/numbers", json={"number": number, **kwargs}))
    def update_number(self, number_id: str, **kwargs: Any) -> PhoneNumber: return PhoneNumber(**self._client.patch(f"{self._base}/numbers/{number_id}", json=kwargs))
    def release_number(self, number_id: str) -> None: self._client.delete(f"{self._base}/numbers/{number_id}")
    def search_available_numbers(self, country: str = "US", **params: Any) -> list[dict]: return self._client.get(f"{self._base}/numbers/available", params={"country": country, **{k: v for k, v in params.items() if v is not None}})
    def validate_number(self, number: str) -> dict: return self._client.post(f"{self._base}/numbers/validate", json={"number": number})
    def get_supported_countries(self) -> list[dict]: return self._client.get(f"{self._base}/numbers/countries")
    def make_call(self, from_number: str, to_number: str, **kwargs: Any) -> Call: return Call(**self._client.post(f"{self._base}/calls", json={"from": from_number, "to": to_number, **kwargs}))
    def get_call(self, call_id: str) -> Call: return Call(**self._client.get(f"{self._base}/calls/{call_id}"))
    def update_call(self, call_id: str, **kwargs: Any) -> Call: return Call(**self._client.patch(f"{self._base}/calls/{call_id}", json=kwargs))
    def end_call(self, call_id: str) -> None: self._client.post(f"{self._base}/calls/{call_id}/end")
    def list_calls(self, **params: Any) -> dict: return self._client.get(f"{self._base}/calls", params={k: v for k, v in params.items() if v is not None})
    def get_recording(self, call_id: str) -> dict: return self._client.get(f"{self._base}/calls/{call_id}/recording")
    def create_conference(self, name: str, **kwargs: Any) -> Conference: return Conference(**self._client.post(f"{self._base}/conferences", json={"name": name, **kwargs}))
    def get_conference(self, conference_id: str) -> Conference: return Conference(**self._client.get(f"{self._base}/conferences/{conference_id}"))
    def end_conference(self, conference_id: str) -> None: self._client.post(f"{self._base}/conferences/{conference_id}/end")
    def list_conferences(self, **params: Any) -> dict: return self._client.get(f"{self._base}/conferences", params={k: v for k, v in params.items() if v is not None})
    def add_conference_participant(self, conference_id: str, call_id: str) -> ConferenceParticipant: return ConferenceParticipant(**self._client.post(f"{self._base}/conferences/{conference_id}/participants", json={"call_id": call_id}))
    def update_conference_participant(self, conference_id: str, participant_id: str, **kwargs: Any) -> ConferenceParticipant: return ConferenceParticipant(**self._client.patch(f"{self._base}/conferences/{conference_id}/participants/{participant_id}", json=kwargs))
    def remove_conference_participant(self, conference_id: str, participant_id: str) -> None: self._client.delete(f"{self._base}/conferences/{conference_id}/participants/{participant_id}")
    def wait_for_call_end(self, call_id: str, poll_interval: float = 2.0, timeout: float = 3600.0) -> Call:
        start = time.time()
        while time.time() - start < timeout:
            c = self.get_call(call_id)
            if c.status in ("completed", "failed", "busy", "no-answer"): return c
            time.sleep(poll_interval)
        raise TimeoutError(f"Call did not end within {timeout}s")
